### PR TITLE
Introduce batchmode

### PR DESCRIPTION
**Pull request by mhinz**
In batchmode you can mark entries without going to their line first.

Use uppercase characters `B`/`S`/`V`/`T` to switch between modes.

E.g. V123 is the same as going to entry 1, hitting <kbd>v</kbd>, then entry 2, hitting <kbd>v</kbd>, then entry 3, hitting <kbd>v</kbd>. Use `<cr>` as usual to open the marked entries.

Using the lowercase variants still works as before and resets an active batchmode.

Implements https://github.com/mhinz/vim-startify/issues/357

---

CC @hupfdule

### CODE REVIEW

**Comment by hupfdule** [comment]
Looks nice,

but there are a few imperfections.

1. When starting batch mode the status line displays e.g. `Batchmode: B` (while it didn't show anything before). When closing batch mode (by pressing `b` on an entry) the status line displays `Batchmode:`. This isn't nice. It does the same when not entering batchmode and directly tagging an entry in non-batchmode. It shouldn't display anything when not in batch mode.

2. While playing around with it it feels a bit strange to use different keys for entering batch mode (e.g. `B`) and for exiting batch mode (e.g. `b`). Especially that I have to select an entry to exit batch mode. How about exiting batch mode by pressing the current batch key again?
So when I am in `B` batch mode, hitting `B` again would exit batch mode.
When I am in `S` batch mode, hitting `S` again would exit batch mode.
When I am in `B` batch mode, hitting `S` would switch the batch mode to S.
Maybe hitting `ESC` would exit any batch mode.

What do you think about it?

**Comment by mhinz** [comment]
Yeah, it's mostly a PoC so far. :-)

But the changes you mentioned are easy to add. Actually, when I wrote it I also expected to see the `Batchmode:`, but didn't. For some reason that only happens with Vim and not with Nvim, which is mainly use. For Nvim, that command-line is cleared on after `Vv`.

And documentation needs to be added, of course.

**Comment by hupfdule** [comment]
Interesting.

With a pure vim, with no plugins than vim-startfiy, The message `Batchmode off` never disappears. Which nvim it disappears after doing some operation.
When actvating all my other plugins, the message stays in nvim, too.

However, it still distracts me a bit. It is nice to have a message that informs me about the disabling of batch mode, but that message should disappear after a small timeout.

However when in Batchmode it should behave differently and never disappear, since the information in which Batchmode I am is important. If nothing is displayed, I am not in any batchmode.

**Comment by hupfdule** [comment]
Hi,

any news on this? Do you intend to change anything regarding the message? Or do you prefer to leave it as it is?

**Comment by mhinz** [comment]
Well, one simply doesn't have that much control over messages.

When in batchmode, you want the message to stay, but any other script could just print over it. Outside of batchmode, you want the message to disappear after a while, but there's no mechanism to that safely. One could abuse a timer function which clears the command-line after 2 seconds or so, but that would clear _any_ message, even when the last message came from a completely different script etc.

Long story short, any solution to this will be error-prone. So.. either we go with the current behaviour or we need to find another way to signalise the current mode that doesn't rely on messages.

But maybe we're overengineering here and should just merge this and get on. :-)

**Comment by hupfdule** [comment]
An alternative I can think of is writing that information directly into the main window like in this screenshot-mockup:

startify

While in batch mode the line `(Batch-selection-mode: B)` is visible. Otherwise it is not.

**Comment by mhinz** [comment]
Well, inserting lines will shift all the other one below, but basically everything in the Startify buffer relies on finding things at certain lines, e.g. mappings and even syntax highlighting.

Honestly, I still think the `:echo` approach, although the most simple is also the correct one.

Let's face it, most people won't even use this new functionality. Many don't even know than you can open multiple buffers at once. When people use batchmode, they will quickly get used to it and use it blindly. They don't switch modes multiple times.

Batchmode is meant to open multiple buffers even quicker. And when the buffers get opened, the message goes away anyway. (Unless people choose to use `shortmess+=F`, but then they have the same "issue" in a lot of other places anyway.)

**Comment by hupfdule** [comment]
Well, you may be right there. It may not be worth the effort. So we could just stick to the `echo` approach. :-)
